local_planner/test/test_star_planner.cpp

```diff
@@ -0,0 +1,99 @@
+#include <gtest/gtest.h>
+
+#include "../src/nodes/star_planner.h"
+#include "../src/nodes/tree_node.h"
+
+using namespace avoidance;
+
+class StarPlannerTests : public ::testing::Test {
+ public:
+  StarPlanner star_planner;
+  float obstacle_half_height = 0.5f;
+  float obstacle_min_x = 0.0f;
+  float obstacle_max_x = 2.5f;
+  float obstacle_y = 2.0f;
+
+  void SetUp() override {
+    ros::Time::init();
+
+    avoidance::LocalPlannerNodeConfig config =
+        avoidance::LocalPlannerNodeConfig::__getDefault__();
+    config.childs_per_node_ = 2;
+    config.n_expanded_nodes_ = 10;
+    star_planner.dynamicReconfigureSetStarParams(config, 1);
+
+    geometry_msgs::PoseStamped position;
+    position.pose.position.x = 1.2f;
+    position.pose.position.y = 0.4f;
+    position.pose.position.z = 4.0f;
+    position.pose.orientation.x = 0.0f;
+    position.pose.orientation.y = 0.0f;
+    position.pose.orientation.z = 0.0f;
+    position.pose.orientation.w = 0.0f;
+
+    geometry_msgs::Point goal;
+    goal.x = 2.0f;
+    goal.y = 14.0f;
+    goal.z = 4.0f;
+
+    pcl::PointCloud<pcl::PointXYZ> cloud;
+    for (float x = obstacle_min_x; x < obstacle_max_x; x += 0.05f) {
+      for (float z = goal.z - obstacle_half_height;
+           z < goal.z + obstacle_half_height; z += 0.05f) {
+        cloud.push_back(pcl::PointXYZ(x, obstacle_y, z));
+      }
+    }
+    costParameters cost_params;
+    const pcl::PointCloud<pcl::PointXYZ> reprojected_points;
+    const std::vector<int> reprojected_points_age;
+
+    star_planner.setParams(cost_params);
+    star_planner.setFOV(270.0, 45.0);
+    star_planner.setReprojectedPoints(reprojected_points,
+                                      reprojected_points_age);
+    star_planner.setPose(position, 0.0);
+    star_planner.setGoal(goal);
+    star_planner.setCloud(cloud);
+  }
+  void TearDown() override {}
+};
+
+TEST_F(StarPlannerTests, buildTree) {
+  // GIVEN: a vehicle position, a goal, and an obstacle in between the stright
+  // line position-goal
+
+  // WHEN: we build the tree
+  star_planner.buildLookAheadTree();
+
+  // THEN: no cell position should be in the obstacle coordinates
+  std::vector<TreeNode> tree_truth;
+  TreeNode n0(0, 0, Eigen::Vector3f(1.2, 0.4, 4));
```

Comment:
Similarly to the other PR, I think we need some other way of making sure this is doing the right thing which doesn't check the exact behavior of the existing implementation, but instead making sure the existing implementation conforms to the API, but will also pass a different implementation conforming to the API. Otherwise we get locked into this implementation, and if we do a rewrite (for whatever reason) the existing tests won't tell us if the new implementation also conforms to the API, but instead that the behavior is identical to the old implementation.

Maybe something like iterating over the points and checking a few things, like that the points move towards the goal, don't get too close to the point cloud, and actually branch outward from the starting pose.